syntax = "proto3";

// Define the message structure for Seller
message RegisterSellerRequest {
    string address = 1;
    string uuid = 2;
}

message RegisterSellerResponse {
    enum Status {
        SUCCESS = 0;
        FAILED = 1;
    }
    Status status = 1;
}

message SellItemRequest {
    string product_name = 1;
    enum Category {
        ELECTRONICS = 0;
        FASHION = 1;
        OTHERS = 2;
    }
    int32 quantity = 2;
    string description = 3;
    string seller_address = 4;
    string seller_uuid = 5;
}

message SellItemResponse {
    enum Status {
        SUCCESS = 0;
        FAILED = 1;
    }
    Status status = 1;
    int32 item_id = 2;
}

message UpdateItemRequest {
    int32 item_id = 1;
    float new_price = 2;
    int32 new_quantity = 3;
    string seller_address = 4;
    string seller_uuid = 5;
}

message UpdateItemResponse {
    enum Status {
        SUCCESS = 0;
        FAILED = 1;
    }
    Status status = 1;
}

message DeleteItemRequest {
    int32 item_id = 1;
    string seller_address = 2;
    string seller_uuid = 3;
}

message DeleteItemResponse {
    enum Status {
        SUCCESS = 0;
        FAILED = 1;
    }
    Status status = 1;
}

message DisplaySellerItemsRequest {
    string seller_address = 1;
    string seller_uuid = 2;
}

message DisplaySellerItemsResponse {
    message Item {
        int32 item_id = 1;
        float price = 2;
        string name = 3;
        enum Category {
            ELECTRONICS = 0;
            FASHION = 1;
            OTHERS = 2;
        }
        string description = 4;
        int32 quantity_remaining = 5;
        string seller_address = 6;
        float rating = 7;
    }
    repeated Item items = 1;
}

// Define the message structure for Buyer
message SearchItemRequest {
    string item_name = 1;
    enum Category {
        ELECTRONICS = 0;
        FASHION = 1;
        OTHERS = 2;
        ANY = 3;
    }
    Category category = 2;
}

message SearchItemResponse {
    message Item {
        int32 item_id = 1;
        float price = 2;
        string name = 3;
        enum Category {
            ELECTRONICS = 0;
            FASHION = 1;
            OTHERS = 2;
        }
        string description = 4;
        int32 quantity_remaining = 5;
        float rating = 6;
        string seller_address = 7;
    }
    repeated Item items = 1;
}

message BuyItemRequest {
    int32 item_id = 1;
    int32 quantity = 2;
    string buyer_address = 3;
}

message BuyItemResponse {
    enum Status {
        SUCCESS = 0;
        FAILED = 1;
    }
    Status status = 1;
}

message AddToWishListRequest {
    int32 item_id = 1;
    string buyer_address = 2;
}

message AddToWishListResponse {
    enum Status {
        SUCCESS = 0;
        FAILED = 1;
    }
    Status status = 1;
}

message RateItemRequest {
    int32 item_id = 1;
    string buyer_address = 2;
    int32 rating = 3;
}

message RateItemResponse {
    enum Status {
        SUCCESS = 0;
        FAILED = 1;
    }
    Status status = 1;
}

// Define the message structure for Notification
message NotifyClient {
    message UpdatedItemDetails {
        int32 item_id = 1;
        float price = 2;
        string name = 3;
        enum Category {
            ELECTRONICS = 0;
            FASHION = 1;
            OTHERS = 2;
        }
        string description = 4;
        int32 quantity_remaining = 5;
        float rating = 6;
        string seller_address = 7;
    }
    UpdatedItemDetails updated_item = 1;
}
